app/Services/AirfieldService.php

```diff
@@ -13,23 +16,168 @@ public function getAllAirfieldsWithRelations() : array
     {
         $airfields = [];
         Airfield::all()->each(function (Airfield $airfield) use (&$airfields) {
-            $prenotePairings = $airfield->prenotePairings()->select(['destination_airfield_id', 'prenote_id'])->get();
-            $prenoteArray = [];
+            $airfieldPairings = $airfield->prenotePairings()->select(['destination_airfield_id', 'prenote_id'])->get();
+            $airfieldArray = [];
 
-            $prenotePairings->each(function (Airfield $airfield) use (&$prenoteArray) {
-                $prenoteArray[$airfield->destination_airfield_id][] = $airfield->prenote_id;
+            $airfieldPairings->each(function (Airfield $airfield) use (&$airfieldArray) {
+                $airfieldArray[$airfield->destination_airfield_id][] = $airfield->prenote_id;
             });
 
             $airfields[] = array_merge(
                 $airfield->toArray(),
                 [
                     'controllers' =>
                     $airfield->controllers()->orderBy('order')->pluck('controller_position_id')->toArray(),
-                    'pairing-prenotes' => $prenoteArray
+                    'pairing-prenotes' => $airfieldArray
                 ]
             );
         });
 
         return $airfields;
     }
+
+    public static function insertIntoOrderBefore(
+        string $airfieldKey,
+        string $positionToInsert,
+        string $insertBefore
+    ): void {
+        try {
+            DB::beginTransaction();
+            // Get the models
+            $insertPosition = ControllerPosition::where('callsign', $positionToInsert)->firstOrFail();
+            $beforePosition = ControllerPosition::where('callsign', $insertBefore)->firstOrFail();
+            $airfield = Airfield::where('code', $airfieldKey)->firstOrFail();
+
+            // Check the insert before is in the handoff order
+            $airfieldBefore = $airfield->controllers->where('id', $beforePosition->id)->first();
+            if (!$airfieldBefore) {
+                throw new OutOfRangeException('Position to insert before not found in top down order');
+            }
+
+            // Increment everything else
+            $controllersToChange = $airfield->controllers()
+                ->wherePivot('order', '>=', $airfieldBefore->pivot->order)
+                ->orderBy('order', 'desc')
+                ->get();
+
+            $controllersToChange->each(function (ControllerPosition $position) use ($airfield) {
+                DB::table('top_downs')
+                    ->where(['controller_position_id' => $position->id, 'airfield_id' => $airfield->id])
+                    ->update(['order' => DB::raw('`order` + 1')]);
+            });
+
+            // Pop the new position in
+            $airfield->controllers()->attach($insertPosition->id, ['order' => $airfieldBefore->pivot->order]);
+            DB::commit();
+        } catch (Exception $exception) {
```

Comment:
I probably know the answer, but is there a more specific exception we can catch here?

Reply:
I guess the behaviour would be the same regardless of the exception flow, whether it be a model not found etc.

Reply:
Logic here is that if anything at all has gone wrong, we probably want to kill the migration so we don't leave the data half-cocked :)

Reply:
All good!